llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -3963,8 +3963,21 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
         Elements.push_back(DAG.getCTSelect(dl, ScalarVT, Tmp1, TVal, FVal, Node->getFlags()));
       }
       Tmp1 = DAG.getBuildVector(VT, dl, Elements);
+    } else if (VT.isFloatingPoint()) {
+      EVT IntegerVT = EVT::getIntegerVT(*DAG.getContext(), VT.getSizeInBits());
+      Tmp2 = DAG.getBitcast(IntegerVT, Tmp2);
+      Tmp3 = DAG.getBitcast(IntegerVT, Tmp3);
+      Tmp1 = DAG.getBitcast(VT, DAG.getCTSelect(dl, IntegerVT, Tmp1, Tmp2, Tmp3, Node->getFlags()));
     } else {
-      Tmp1 = DAG.getCTSelect(dl, Tmp1.getValueType(), Tmp1, Tmp2, Tmp3);
+      assert(VT.isInteger());
+      EVT HalfVT = VT.getHalfSizedIntegerVT(*DAG.getContext());
+      SDValue Tmp2Lo, Tmp2Hi;
+      SDValue Tmp3Lo, Tmp3Hi;
+      std::tie(Tmp2Lo, Tmp2Hi) = DAG.SplitScalar(Tmp2, dl, HalfVT, HalfVT);
+      std::tie(Tmp3Lo, Tmp3Hi) = DAG.SplitScalar(Tmp3, dl, HalfVT, HalfVT);
```

Comment:
Since LLVM uses C++17, maybe we should use structure bindings instead? 

So something like this:
```suggestion
      auto [Tmp3Lo, Tmp3Hi] = DAG.SplitScalar(Tmp3, dl, HalfVT, HalfVT);
```